bandit/bandit.go

```diff
@@ -353,6 +363,9 @@ type Dialer interface {
 	// connections created via this dialer.
 	DataRecv() uint64
 
+	// IsReady indicates when the dialer is ready for dialing
+	IsReady() (bool, error)
```

Comment:
There's a lot of complexity (in the bandit and bypass code) created by two things:
1. The double return value here.  The first value and second value each need to be checked independently before calling code can proceed to use the dialer.
2. A lack of synchronization provided by this function.  Callers are left having to poll this function.  Since we're trying to communicate when an event happens (when the dialer is ready), this feels like a natural use case for a channel.

Maybe instead, something like:
```
// Ready returns a channel which will have a value on it only when initialization
// of the dialer is complete. If initialization failed, the channel will have a non-nil
// error value.
Ready() chan<- error
```

Then callers use this like:
```
select {
case err := <-dialer.Ready():
    if err != nil {
        return fmt.Errorf("dialer initialization failed: %w", err)
    }
case <-ctx.Done():
    return ctx.Error()
}
// Here we know the dialer is ready to use.
```

Your way definitely works, to be clear!  I'm fine with that if you prefer it.

Reply:
Hmmmmm this is valid and would work perfectly on the bandit side, but on bypass, how do we know when we should start the proxy asynchronously? Should we start all proxies asynchronously?

Reply:
hmmmm I'll return nil by default, if the chan is nil It's ready by default